src/cargo/ops/cargo_new.rs

```diff
@@ -412,74 +423,123 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
         (None, None, name, None) => name,
     };
 
-    let mut cargotoml_path_specifier = String::new();
-
-    // Calculare what [lib] and [[bin]]s do we need to append to Cargo.toml
-
-    for i in &opts.source_files {
-        if i.bin {
-            if i.relative_path != "src/main.rs" {
-                cargotoml_path_specifier.push_str(&format!(r#"
-[[bin]]
-name = "{}"
-path = {}
-"#, i.target_name, toml::Value::String(i.relative_path.clone())));
-            }
-        } else {
-            if i.relative_path != "src/lib.rs" {
-                cargotoml_path_specifier.push_str(&format!(r#"
-[lib]
-name = "{}"
-path = {}
-"#, i.target_name, toml::Value::String(i.relative_path.clone())));
-            }
-        }
+    let templates_dir = config.template_path();
+    if fs::metadata(&templates_dir).is_err() {
+        try!(fs::create_dir(&templates_dir));
     }
+   // create lib & bin templates if not already present.
+    let lib_template = templates_dir.join("lib");
+    let bin_template = templates_dir.join("bin");
+    try!(create_bin_template(&bin_template));
+    try!(create_lib_template(&lib_template));
+
+    let template_dir = match opts.template {
+        // given template is a remote git repository & needs to be cloned
+        // This will be cloned to .cargo/templates/<repo_name> where <repo_name>
+        // is the last component of the given URL. For example:
+        //
+        //      http://github.com/rust-lang/some-template
+        //      <repo_name> = some-template
+        Some(template) if template.starts_with("http") ||
+                          template.starts_with("git@") => {
+            let path = PathBuf::from(template);
+
+            let repo_name = match path.components().last().unwrap() {
+                Component::Normal(p) => p,
+                _ => {
+                    return Err(human(format!(
+                        "Could not determine repository name from: {}", path.display())))
+                }
+            };
+            let template_dir = templates_dir.join(repo_name);
+
+            match Repository::clone(template, &*template_dir) {
```

Comment:
Cargo typically issues a status message for operations that may take a long time, so perhaps this could indicate that a repository is being cloned?

I think this may also wish to use the helpers in `git/utils.rs` to ensure this handles pieces such as authentication, retries, etc.